src/logic/safe/store/actions/fetchSafe.ts

```diff
@@ -1,178 +1,140 @@
-import GnosisSafeSol from '@gnosis.pm/safe-contracts/build/contracts/GnosisSafe.json'
-import { List, Set, Map } from 'immutable'
-import { Action, Dispatch } from 'redux'
-import { AbiItem } from 'web3-utils'
+import { List } from 'immutable'
+import { Dispatch } from 'redux'
 
-import generateBatchRequests from 'src/logic/contracts/generateBatchRequests'
-import { getLocalSafe, getSafeName } from 'src/logic/safe/utils'
-import { enabledFeatures, safeNeedsUpdate } from 'src/logic/safe/utils/safeVersion'
-import { sameAddress } from 'src/logic/wallets/ethAddresses'
-import { getBalanceInEtherOf } from 'src/logic/wallets/getWeb3'
-import { addSafeOwner } from 'src/logic/safe/store/actions/addSafeOwner'
-import { removeSafeOwner } from 'src/logic/safe/store/actions/removeSafeOwner'
-import updateSafe from 'src/logic/safe/store/actions/updateSafe'
+import { addressBookSelector } from 'src/logic/addressBook/store/selectors'
+import { updateSafe } from 'src/logic/safe/store/actions/updateSafe'
 import { makeOwner } from 'src/logic/safe/store/models/owner'
-import { checksumAddress } from 'src/utils/checksumAddress'
-import { SafeOwner, SafeRecordProps } from 'src/logic/safe/store/models/safe'
-import { AppReduxState } from 'src/store'
-import { latestMasterContractVersionSelector, safeTotalFiatBalanceSelector } from 'src/logic/safe/store/selectors'
-import { getSafeInfo } from 'src/logic/safe/utils/safeInformation'
+import { ModulePair, SafeOwner, SafeRecordProps, SpendingLimit } from 'src/logic/safe/store/models/safe'
+import { getLocalSafe } from 'src/logic/safe/utils'
 import { getModules } from 'src/logic/safe/utils/modules'
+import { allSettled } from 'src/logic/safe/utils/allSettled'
+import { getSafeInfo, SafeInfo } from 'src/logic/safe/utils/safeInformation'
+import { enabledFeatures, safeNeedsUpdate } from 'src/logic/safe/utils/safeVersion'
 import { getSpendingLimits } from 'src/logic/safe/utils/spendingLimits'
-import { LOADED_SAFE_KEY } from 'src/utils/constants'
+import { sameAddress } from 'src/logic/wallets/ethAddresses'
+import { AppReduxState } from 'src/store'
+import { checksumAddress } from 'src/utils/checksumAddress'
+
+/**
+ * Recovers Safe's remote information along with its modules and spendingLimits if there's any
+ * @param {SafeInfo} remoteSafeInfo
+ * @param {string} checksummedSafeAddress
+ * @returns Promise<Partial<SafeRecordProps>>
+ */
+const extractRemoteSafeInfo = async (
+  remoteSafeInfo: SafeInfo,
+  checksummedSafeAddress: string,
+): Promise<Partial<SafeRecordProps>> => {
+  const safeInfo: Partial<SafeRecordProps> = {
+    spendingLimits: undefined,
+    modules: undefined,
+  }
+  const safeInfoModules = remoteSafeInfo.modules.map(({ value }) => value)
 
-const buildOwnersFrom = (safeOwners: string[], localSafe?: SafeRecordProps): List<SafeOwner> => {
-  const ownersList = safeOwners.map((ownerAddress) => {
-    const convertedAdd = checksumAddress(ownerAddress)
+  if (safeInfoModules) {
```

Comment:
This will always be true. I think we need to check the length of the array.

Reply:
:grimacing: